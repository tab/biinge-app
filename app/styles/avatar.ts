import { StyleSheet } from "react-native"

export const avatarStyles = StyleSheet.create({
  root: {
    overflow: "hidden",
  },
  image: {
    borderRadius: 50,
    height: "100%",
    width: "100%",
  },
})
