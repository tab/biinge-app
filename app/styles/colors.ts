const colors = {
  americanSilver: "#d1cecf",
  black: "#181818",
  darkBlue: "#2a3642",
  darkCandyAppleRed: "#9f0909",
  graniteGray: "#666666",
  gray: "#b5b5b5",
  grayDark: "#6e6969",
  gunmetal: "#2b2835",
  lotion: "#fafafa",
  orangeYellow: "#eec01e",
  palmLeaf: "#699d3a",
  raisinBlack: "#221f1f",
  white: "#ffffff",
}

export default colors
