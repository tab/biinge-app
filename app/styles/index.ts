export * from "styles/layout"
export * from "styles/avatar"
export * from "styles/badge"
export * from "styles/button"
export * from "styles/flash"
export * from "styles/form"
export * from "styles/loading"

export * from "styles/login"
export * from "styles/movie"
export * from "styles/search"
export * from "styles/profile"
