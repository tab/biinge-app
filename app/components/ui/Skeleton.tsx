import * as React from "react"
import {
  View,
  StyleSheet,
  ViewStyle,
  Dimensions,
  LayoutRectangle,
} from "react-native"
import Animated, {
  Easing,
  interpolate,
  useAnimatedStyle,
  useSharedValue,
  withRepeat,
  withTiming,
} from "react-native-reanimated"
import MaskedView from "@react-native-masked-view/masked-view"
import LinearGradient from "react-native-linear-gradient"
import { useTheme } from "@react-navigation/native"

import { darkTheme, lightTheme } from "styles/theme"

const SCREEN_WIDTH = Dimensions.get("window").width

interface SkeletonProps {
  children: React.JSX.Element | React.JSX.Element[]
  speed?: number
  direction?: "left" | "right"
}

export default function Skeleton({
  children,
  speed = 800,
  direction = "right",
}: SkeletonProps) {
  const { dark } = useTheme()

  const backgroundColor = dark
    ? darkTheme.colors.card
    : lightTheme.colors.border
  const highlightColor = dark
    ? darkTheme.colors.background
    : lightTheme.colors.card

  const [layout, setLayout] = React.useState<LayoutRectangle>()
  const animated = useSharedValue(0)

  React.useEffect(() => {
    animated.value = withRepeat(
      withTiming(1, {
        duration: speed,
        easing: Easing.ease,
      }),
      -1,
    )
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
  const animatedStyle = useAnimatedStyle(() => {
    const translateX = interpolate(
      animated.value,
      [0, 1],
      direction === "right"
        ? [-SCREEN_WIDTH, SCREEN_WIDTH]
        : [SCREEN_WIDTH, -SCREEN_WIDTH],
    )
    return {
      transform: [
        {
          translateX: translateX,
        },
      ],
    }
  })
  const viewStyle = React.useMemo<ViewStyle>(
    () => ({ backgroundColor, overflow: "hidden" }),
    [backgroundColor],
  )

  const getChildren = React.useCallback(
    (element: React.JSX.Element | React.JSX.Element[]) => {
      return React.Children.map(
        element,
        (child: React.JSX.Element, index: number) => {
          let style: ViewStyle
          if (child.type.displayName === "SkeletonItem") {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars,@typescript-eslint/no-shadow
            const { children, ...styles } = child.props
            style = styles
          } else {
            style = child.props.style
          }
          if (child.props.children) {
            return (
              <View key={index} style={style}>
                {getChildren(child.props.children)}
              </View>
            )
          } else {
            return (
              <View key={index} style={styles.childContainer}>
                <View style={[style, viewStyle]} />
              </View>
            )
          }
        },
      )
    },
    [viewStyle],
  )

  return layout?.width && layout?.height ? (
    <MaskedView
      style={{ height: layout.height, width: layout.width }}
      maskElement={
        <View style={styles.transparent}>{getChildren(children)}</View>
      }
    >
      {/* eslint-disable-next-line react-native/no-inline-styles */}
      <View style={{ flexGrow: 1, backgroundColor }} />
      {speed > 0 && (
        <Animated.View
          style={[
            // eslint-disable-next-line react-native/no-inline-styles
            {
              flexDirection: "row",
            },
            StyleSheet.absoluteFillObject,
            animatedStyle,
          ]}
        >
          <MaskedView
            style={StyleSheet.absoluteFill}
            maskElement={
              <LinearGradient
                start={{ x: 0, y: 0 }}
                end={{ x: 1, y: 0 }}
                style={[StyleSheet.absoluteFill]}
                colors={["transparent", "black", "transparent"]}
              />
            }
          >
            <View
              style={[
                StyleSheet.absoluteFill,
                { backgroundColor: highlightColor },
              ]}
            />
          </MaskedView>
        </Animated.View>
      )}
    </MaskedView>
  ) : (
    <View
      onLayout={(event) => {
        setLayout(event.nativeEvent.layout)
      }}
    >
      {getChildren(children)}
    </View>
  )
}

interface SkeletonItem extends ViewStyle {
  children?: React.JSX.Element | React.JSX.Element[]
}

Skeleton.Item = ({ children, ...style }: SkeletonItem): React.JSX.Element => (
  <View style={style}>{children}</View>
)

// @ts-ignore
Skeleton.Item.displayName = "SkeletonItem"

const styles = StyleSheet.create({
  childContainer: {
    position: "relative",
  },
  transparent: {
    backgroundColor: "transparent",
  },
  gradient: {
    flex: 1,
  },
})
