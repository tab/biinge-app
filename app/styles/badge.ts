import { StyleSheet } from "react-native"
import colors from "styles/colors"

export const badgeStyles = StyleSheet.create({
  container: {
    backgroundColor: "rgba(25, 25, 25, 0.25)",
    borderRadius: 50,
    paddingVertical: 0,
    paddingBottom: 2,
    paddingHorizontal: 10,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    position: "absolute",
    top: 22,
    right: 18,
  },
  text: {
    color: colors.white,
    fontSize: 12,
    fontWeight: "400",
    lineHeight: 21,
  },
})
