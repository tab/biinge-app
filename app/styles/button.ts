import { StyleSheet } from "react-native"
import colors from "styles/colors"

export const buttonStyles = StyleSheet.create({
  button: {
    alignItems: "center",
    backgroundColor: colors.gunmetal,
    borderRadius: 50,
    justifyContent: "center",
    padding: 15,
  },
  title: {
    color: colors.white,
    fontWeight: "700",
    fontSize: 14,
    lineHeight: 16,
  },
  text: {
    color: colors.white,
    textAlign: "center",
    fontSize: 18,
    fontWeight: "bold",
  },
})

export const activeButtonStyles = StyleSheet.create({
  button: {
    ...buttonStyles.button,
    backgroundColor: colors.orangeYellow,
    width: "100%",
  },
  title: {
    ...buttonStyles.title,
    color: colors.white,
  },
})
