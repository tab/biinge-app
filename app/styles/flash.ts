import { StyleSheet } from "react-native"
import colors from "styles/colors"

export const flashStyles = StyleSheet.create({
  root: {
    backgroundColor: "transparent",
    height: 50,
    width: "100%",
  },
  text: {},
  title: {
    textAlign: "center",
    fontSize: 16,
    fontWeight: "500",
    color: colors.white,
  },
})
