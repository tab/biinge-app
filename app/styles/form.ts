import { StyleSheet } from "react-native"

import { darkTheme, lightTheme } from "styles/theme"
import colors from "styles/colors"

export const inputStyles = StyleSheet.create({
  root: {
    color: colors.black,
    fontSize: 16,
    borderWidth: 1,
    padding: 12,
    backgroundColor: colors.white,
    borderRadius: 4,
    borderColor: colors.black,
  },
  dark: {
    backgroundColor: darkTheme.colors.card,
    color: darkTheme.colors.text,
  },
  light: {
    backgroundColor: lightTheme.colors.card,
    color: lightTheme.colors.text,
  },
  text: {
    color: colors.white,
    textAlign: "center",
    fontSize: 17,
    fontWeight: "bold",
  },
})

export const inputErrorStyles = StyleSheet.create({
  container: {
    alignItems: "flex-start",
    marginTop: 5,
  },
  text: {
    color: colors.white,
    textAlign: "center",
    fontSize: 15,
  },
})

export const formErrorStyles = StyleSheet.create({
  container: {
    marginBottom: 15,
  },
  text: {
    color: colors.darkCandyAppleRed,
    textAlign: "center",
    fontSize: 15,
  },
})
