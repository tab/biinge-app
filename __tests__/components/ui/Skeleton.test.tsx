import React from "react"
import { View } from "react-native"
import { render } from "@testing-library/react-native"
import { it } from "@jest/globals"

import Skeleton from "components/ui/Skeleton"
import { horizontalStubListStyles } from "styles"

describe("Skeleton", () => {
  it("renders correctly", () => {
    const tree = render(
      <Skeleton>
        <View style={horizontalStubListStyles.root}>
          <View style={horizontalStubListStyles.itemMovie} />
          <View style={horizontalStubListStyles.itemMovie} />
          <View style={horizontalStubListStyles.itemMovie} />
        </View>
      </Skeleton>,
    )
    expect(tree).toMatchSnapshot()
  })
})
