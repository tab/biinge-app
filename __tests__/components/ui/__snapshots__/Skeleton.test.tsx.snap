// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton renders correctly 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "paddingLeft": 10,
      }
    }
  >
    <View
      style={
        {
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 8,
              "height": 374.9625,
              "marginRight": 10,
              "width": 249.975,
            },
            {
              "backgroundColor": "#d1cecf",
              "overflow": "hidden",
            },
          ]
        }
      />
    </View>
    <View
      style={
        {
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 8,
              "height": 374.9625,
              "marginRight": 10,
              "width": 249.975,
            },
            {
              "backgroundColor": "#d1cecf",
              "overflow": "hidden",
            },
          ]
        }
      />
    </View>
    <View
      style={
        {
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 8,
              "height": 374.9625,
              "marginRight": 10,
              "width": 249.975,
            },
            {
              "backgroundColor": "#d1cecf",
              "overflow": "hidden",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
